### ISSUE TITLE
Constructing Empty Polygon Creates Empty Geometry Collection, or something else...

### ISSUE THREAD

**Issue by bixb0012** [Toblerity/Shapely#338]
Looking at issue #332 got me stuck down another rabbit hole.  I haven't had time to fully look into it, but something is off with constructing empty geometries.
```python
>>> from shapely.geometry import Polygon, GeometryCollection
>>> from shapely.wkt import loads
>>> empty_poly_wkt = loads('POLYGON EMPTY')
>>> empty_poly_wkt.wkt
'POLYGON EMPTY'
>>> isinstance(empty_poly_wkt, Polygon)
True
>>> isinstance(empty_poly_wkt, GeometryCollection)
False
>>> 
>>> empty_poly_none = Polygon()  #or Polygon(None)
>>> empty_poly_none.wkt
'GEOMETRYCOLLECTION EMPTY'
>>> isinstance(empty_poly_none, Polygon)
True
>>> isinstance(empty_poly_none, GeometryCollection)
False
>>> 
>>> empty_poly_wkt.wkb
'\x01\x03\x00\x00\x80\x01\x00\x00\x00\x00\x00\x00\x00'
>>> empty_poly_none.wkb
'\x01\x07\x00\x00\x00\x00\x00\x00\x00'
>>> 
```
It is clear that constructing an empty geometry from well-known text creates a different geometry than constructing an empty geometry using the default class constructor.  The results from well-known text seem correct while results from the default class constructor seems off, i.e., an empty polygon shouldn't have a WKT of empty geometry collection.

**Comment by sgillies** [comment]
@bixb0012 Shapely doesn't currently distinguish between a Polygon that has an empty set of points and any other kind of geometry with an empty set of points. To me, the distinction isn't meaningful. If you have no points – nothing – what's the useful distinction between a point-shaped nothingness and a polygon-shaped nothingness?

I settled on `'GEOMETRYCOLLECTION EMPTY'` as the most generic nothingness, but I wouldn't be opposed to a PR that had more specifically typed nothingnesses that better matched other programs.

**Comment by bixb0012** [comment]
Regardless of the broader philosophical debate/question over emptiness and empty geometries, defaulting to empty geometry collections for all empty geometries appears to create inconsistencies or discrepancies between certain functions/checks.

If I can find some time over this holiday season, I may take you up on that PR suggestion.

**Comment by snorfalorpagus** [comment]
I'm reopening this as I've done some exploration of the issue and I think it's worth further discussion.

It took me a little while to work out where the geometries pick up `GEOMETRYCOLLECTION EMPTY` as a default. It's done in `BaseGeometry`:

https://github.com/Toblerity/Shapely/blob/1f5c11918a3d26d17aaade63da3e1bd6d5b348db/shapely/geometry/base.py#L168
https://github.com/Toblerity/Shapely/blob/1f5c11918a3d26d17aaade63da3e1bd6d5b348db/shapely/geometry/base.py#L195

I figured it would be easy enough to add a couple of lines to the `__init__` method of each geometry type to default to a different geometry. However, this highlighted an issue with WKB and the `POINT EMPTY` geometry.

See links below for an explanation of the issue. In summary, there isn't really a WKB representation of `POINT EMPTY`.

http://gis.stackexchange.com/questions/106932/empty-geometries-in-geometrycollection
http://mateusz.loskot.net/posts/2010/02/26/sqlgeometry-and-point-empty-in-wkb/
http://postgis.17.x6.nabble.com/POINT-EMPTY-in-WKB-td3592406.html

This actually creates an issue when trying to pickle such a geometry in shapely. Consider the following:

```
>>> import pickle
>>> import shapely.wkt
>>> p = shapely.wkt.loads('POINT EMPTY')
>>> p.wkt
'POINT EMPTY'
>>> p.wkb
b''
>>> pickle.loads(pickle.dumps(p))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/snorf/Desktop/Shapely/shapely/geometry/base.py", line 233, in __setstate__
    self.__geom__ = deserialize_wkb(state)
  File "/Users/snorf/Desktop/Shapely/shapely/geometry/base.py", line 109, in deserialize_wkb
    "Could not create geometry because of errors while reading input.")
shapely.geos.ReadingError: Could not create geometry because of errors while reading input.
```

An empty point geometry doesn't occur naturally in shapely. I can't think of an operating that would return it, so the only way to create one is using `wkt.loads` as above. If we moved to creating empty points I'm sure this would become an issue for users. One approach mentioned in the 2nd link is to use `MULTIPOINT EMPTY` for this particular case. The other point is that WKT/WKB doesn't have a specific representation for a LinearRing. Perhaps `GEOMETRYCOLLECTION EMPTY` isn't so bad after all (provided that it's explained in the manual).

**Comment by snorfalorpagus** [comment]
More trouble with empty geometries:

```
>>> from shapely.geometry import Polygon
>>> p1 = Polygon()
>>> p2 = Polygon()
>>> p1 == p2
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/snorf/Desktop/Shapely/shapely/geometry/polygon.py", line 255, in __eq__
    tuple(self.exterior.coords),
AttributeError: 'NoneType' object has no attribute 'coords'
```

**Comment by snorfalorpagus** [comment]
The issue with empty polygons was resolved in 8883cb69f31fcf1b83bb5240cb175163c0056b2b.

Closing this issue again - see comments in #352.
